quinn/src/platform/unix.rs

```diff
@@ -373,3 +399,40 @@ pub const BATCH_SIZE: usize = 32;
 
 #[cfg(any(target_os = "macos", target_os = "ios"))]
 pub const BATCH_SIZE: usize = 1;
+
+#[cfg(target_os = "linux")]
+lazy_static! {
+    static ref CAPABILITIES: super::UdpCapabilities = {
+        /// Checks whether GSO support is available by setting the UDP_SEGMENT
+        // option on a socket.
+        fn supports_gso() -> bool {
+            const GSO_SIZE: libc::c_int = 1500;
+
+            let socket = match std::net::UdpSocket::bind("[::]:0") {
+                Ok(socket) => socket,
+                Err(_) => return false,
+            };
+
+            let rc = unsafe {
+                libc::setsockopt(
+                    socket.as_raw_fd(),
+                    libc::SOL_UDP,
+                    libc::UDP_SEGMENT,
+                    &GSO_SIZE as *const _ as _,
+                    mem::size_of_val(&GSO_SIZE) as _,
+                )
+            };
+
+            rc != -1
+        }
+
+        super::UdpCapabilities {
+            gso: supports_gso(),
+        }
+    };
```

Comment:
Could we do this more compactly, like this?

```rust
static ref CAPABILITIES: UdpCapabilities = UdpCapabilities {
    gso: match std::net::UdpSocket::bind("[::]:0") {
        Err(_) => false,
        Ok(socket) => {
             const GSO_SIZE: libc::c_int = 1500;
             unsafe {
                libc::setsockopt(
                    socket.as_raw_fd(),
                    libc::SOL_UDP,
                    libc::UDP_SEGMENT,
                    &GSO_SIZE as *const _ as _,
                    mem::size_of_val(&GSO_SIZE) as _,
                )
            } != -1
        }
    },
};
```

Personally I'm not really a fan of nested functions.

Reply:
I don't like this one. If we add more capabilities, then this will get a super long block. Imagine having 3 unrelated capabilities here. And the right drift due to the lack of early returns also doesn't make it nicer.

I'm fine with making it either a top level function or move it into a `mod linux`. The strongest reason it is a nested function at the moment is that it avoids requiring another `cfg(target_os)` block, since it is already guarded by the surrounding one.

Reply:
I think a `mod linux` would be better, which should then also contain `set_segment_size()`.

Reply:
(Sorry for the slow turnaround on this.)

Reply:
Moved it into `mod linux`. Not sure about `set_segment_size`. We would then also a `mod not_linux`, and be able to dispatch on the same name. I feel like this is more weird than worthwhile.

Reply:
What if we give the module a name that's more about the contained functionality? Then it would be pretty natural, I think? Like maybe `mod gso` or `mod segment` (with `set_size()` and `is_supported()` in it), with one definition for Linux and one definition for other platforms.

Reply:
I updated for this. Looks reasonable to me.